llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp

```diff
@@ -0,0 +1,131 @@
+        Func.getFunctionType()->getNumParams() == 0) {
+      EmptyFuncs.push_back(&Func);
+    }
+    for (llvm::BasicBlock &Block : Func) {
+      for (llvm::Instruction &Instr : Block) {
+        if (llvm::CallInst *CallInstruction =
+                llvm::dyn_cast<llvm::CallInst>(&Instr)) {
+          llvm::Function *Called = CallInstruction->getCalledFunction();
+          Called = findEmptyFunction(Called);
+          if (!Called || Called == &Func) {
+            continue;
+          }
+          std::map<llvm::BasicBlock *, llvm::BasicBlock *> BlockMap;
+          llvm::ValueToValueMapTy VMap;
+          llvm::BasicBlock *SplittedBlock = splitBlockBefore(&Block, &Instr);
+          llvm::BasicBlock *NextBlockPlace = SplittedBlock->getNextNode();
+          for (llvm::BasicBlock &EmptyBasicBlock : *Called) {
+            llvm::BasicBlock *TmpBlock =
+                llvm::BasicBlock::Create(CTX, "", &Func, NextBlockPlace);
+            BlockMap[&EmptyBasicBlock] = TmpBlock;
+            NextBlockPlace = TmpBlock->getNextNode();
+          }
+          for (llvm::BasicBlock &EmptyBasicBlock : *Called) {
+            for (llvm::Instruction &EmptyInstr : EmptyBasicBlock) {
+              llvm::BasicBlock *CurrentBlock = BlockMap[&EmptyBasicBlock];
+              if (llvm::isa<llvm::ReturnInst>(&EmptyInstr)) {
+                llvm::BranchInst::Create(&Block)->insertInto(
+                    CurrentBlock, CurrentBlock->end());
+              } else {
+                llvm::Instruction *NewInstr = EmptyInstr.clone();
+                if (llvm::isa<llvm::BranchInst>(*NewInstr)) {
+                  unsigned OpNum = NewInstr->getNumOperands();
+                  for (unsigned Ind = 0; Ind < OpNum; Ind++) {
+                    llvm::Value *CurOperand = NewInstr->getOperand(Ind);
+                    if (llvm::isa<llvm::BasicBlock>(*CurOperand)) {
+                      NewInstr->setOperand(
+                          Ind, llvm::dyn_cast<llvm::Value>(
+                                   BlockMap[llvm::dyn_cast<llvm::BasicBlock>(
+                                       CurOperand)]));
+                    }
+                  }
+                }
+                NewInstr->insertInto(CurrentBlock, CurrentBlock->end());
+                llvm::RemapInstruction(NewInstr, VMap,
+                                       llvm::RF_NoModuleLevelChanges |
+                                           llvm::RF_IgnoreMissingLocals);
+                VMap[&EmptyInstr] = NewInstr;
+              }
+            }
+          }
+          SplittedBlock->getTerminator()->setOperand(
+              0,
+              llvm::dyn_cast<llvm::Value>(BlockMap[&Called->getEntryBlock()]));
+          InstrDeleteList.push_back(&Instr);
+        }
+      }
+    }
+    for (llvm::Instruction *Instr : InstrDeleteList) {
+      Instr->eraseFromParent();
+    }
+    return llvm::PreservedAnalyses::all();
+  }
+
+private:
+  std::vector<llvm::Function *> EmptyFuncs;
```

Comment:
I'd suggest to simplify a bit the pass in the following way:
It just searches for `call` instructions in a processed function and if called function doesn't have arguments and doesn't return anything then inline it after `call`?